tools/ci_connector_ops/ci_connector_ops/pipelines/publish.py

```diff
@@ -0,0 +1,232 @@
+
+
+class BuildConnectorForPublish(Step):
+
+    title = "Build connector for publish"
+
+    async def _run(self) -> Tuple[StepResult, List[Container]]:
+        build_connector_results_per_platform = await builds.run_connector_build(self.context)
+        build_connector_results = [build_result for build_result, _ in build_connector_results_per_platform.values()]
+        built_connectors = [built_connector for _, built_connector in build_connector_results_per_platform.values()]
+
+        if not all([build_result.status is StepStatus.SUCCESS for build_result in build_connector_results]):
+            return StepResult(self, status=StepStatus.FAILURE), []
+
+        return StepResult(self, status=StepStatus.SUCCESS), built_connectors
+
+
+class PushConnectorImageToRegistry(PublishStep):
+    title = "Push connector image to registry"
+
+    @property
+    def latest_docker_image_name(self):
+        return f"{self.context.metadata['dockerRepository']}:latest"
+
+    async def _run(self, built_containers_per_platform: List[Container]) -> StepResult:
+        try:
+            image_ref = await built_containers_per_platform[0].publish(
+                f"docker.io/{self.docker_image_name}", platform_variants=built_containers_per_platform[1:]
+            )
+            if not self.pre_release:
+                image_ref = await built_containers_per_platform[0].publish(
+                    f"docker.io/{self.latest_docker_image_name}", platform_variants=built_containers_per_platform[1:]
+                )
+            return StepResult(self, status=StepStatus.SUCCESS, stdout=f"Published {image_ref}")
+        except QueryError as e:
+            return StepResult(self, status=StepStatus.FAILURE, stderr=str(e))
+
+
+class InvalidSpecOutputError(Exception):
+    pass
+
+
+class UploadSpecToCache(PublishStep):
+    title = "Upload connector spec to spec cache bucket"
+    default_spec_file_name = "spec.json"
+    cloud_spec_file_name = "spec.cloud.json"
+
+    def _parse_spec_output(self, spec_output: str) -> str:
+        for line in spec_output.split("\n"):
+            try:
+                parsed_json = json.loads(line)
+                if parsed_json["type"] == "SPEC":
+                    return json.dumps(parsed_json)
+            except (json.JSONDecodeError, KeyError):
+                continue
+        raise InvalidSpecOutputError("Could not parse the output of the spec command.")
+
+    async def _get_connector_spec(self, connector: Container, deployment_mode: str) -> str:
+        spec_output = await connector.with_env_variable("DEPLOYMENT_MODE", deployment_mode).with_exec(["spec"]).stdout()
+        return self._parse_spec_output(spec_output)
+
+    def _get_spec_as_file(self, spec: str, name="spec_to_cache.json") -> File:
+        return self.context.get_connector_dir().with_new_file(name, spec).file(name)
+
+    @property
+    def spec_key_prefix(self):
```

Comment:
Can we move this to library/location where more of our app has access to the naming logic?

The reason I ask is that we have a few services that have to access the spec cache. Would be great to start standardizing that.

Let me know if thats too big of a task and needs to be a separate issue

Reply:
I'd indeed prefer to do it in a separate issue. Probably when we'll move the pipeline code to `airbyte-ci`. It'll be easier to share code between metadata and connectors CI/CD. I logged an issue: #25419